src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs

```diff
@@ -104,8 +91,9 @@ internal int ReadByte()
             {
                 if (_decryptedBytesCount > 0)
                 {
-                    int b = InternalBuffer[_decryptedBytesOffset++];
+                    int b = _internalBuffer[_decryptedBytesOffset++];
                     _decryptedBytesCount--;
+                    ReturnReadBufferIfEmpty();
```

Comment:
This seems like a significant change in behavior.  Why are we doing this in a change about removing PinnableBufferCache?  Do we really understand the impact of this change?

Reply:
Is this read buffer the same byte[] that's passed to NetworkStream under SslStream?  If so, there is the potential concern that we'd be forced to re-pin on every operation, if we end up getting back a different buffer from the pool, whereas if we hold onto the same buffer, the SocketAsyncEventArgs optimizations will kick in and we won't need to.

Reply:
@Drawaes, thinking about this more, I am concerned about the read buffer change.  Do any of the tests you've run validate perf of SslStream over a NetworkStream?

Reply:
I am not 100% sure about what you mean here? Am I understanding it correctly that you are saying

If I call
```
await stream.WriteAsync(someData);
// do some other stuff
await stream.WriteAsync(someData);
```

That from the first write to the second write socket has held a pin on my someData buffer?

Reply:
> That from the first write to the second write socket has held a pin on my someData buffer?

The way things are currently implemented... yes.

Reply:
So what happens if I call
` stream.WriteAsync(bla); `
Then don't call again for some prolonged period, is my buffer no longer eligible for GC and also causing memory fragmentation? (outside of the SslStream case here) ?

Reply:
> is my buffer no longer eligible for GC and also causing memory fragmentation

It's pinned, and everything that comes with it.

Reply:
It's actually interesting. Not sure I like the behaviour of the pin effects leaking, but putting that aside for a second and really unrelated to this PR. 

With buffers, you can pin them, I was thinking that an API for stream not just to pass in a buffer as is currently suggested, but where you could pass in a pin from the buffer as it has such a thing might provide a perf bump. So say SslStream could pin to interop and do the encrypt, and pass the pin to the underlying stream (network or otherwise) and release it after the operation has finished. This could just be ignored if the underlying stream doesn't support it with no harm but if it could use it then there could be a chance to reduce pin/free's all the way down the chain. 

Which btw pipelines avoids in a way by creating slabs of prepinned buffers.

Reply:
> I was thinking that an API for stream not just to pass in a buffer as is currently suggested, but where you could pass in a pin from the buffer as it has such a thing might provide a perf bump

With https://github.com/dotnet/corefx/issues/22608, you'll be able to give a `Buffer<T>` (now called `Memory<T>`) to SocketAsyncEventArgs.  Calling `Retain` on an already pinned buffer should be negligable cost.  If you want to use a pool of already pinned memory, you can.

Reply:
Regardless, let's circle back to my concern.  Have we validated the impact of this on SslStream over a NetworkStream?  Separately it's worthwhile re-evaluating the long-standing behavior of pinning in SocketAsyncEventArgs.SetBuffer so that individual send/receive operations don't have to, but that is the current behavior, and we should validate we're not regressing in the face of it.

Reply:
Agreed, I was unaware of that behaviour and will look into the impact.

Reply:
Thanks